src/viewer.js

```diff
@@ -143,6 +143,14 @@ export class Viewer {
     /** @const @private {!Object<string, string>} */
     this.params_ = {};
 
+    /** @private {?function(*)} */
+    this.onVisibilityResolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.onVisibilityPromise_ = new Promise((resolve) => {
+      this.onVisibilityResolve_ = resolve;
```

Comment:
No, you can resolve the promise right here. Start with the actual property in this class. You want to have a promise `Viewer.whenVisible`. And then the question is when this promise will resolve. The answer is: it may need to be resolved immediately if the document already visible or it may become visible later based on two callbacks via document-state and viewer. 

Whether or not the document is fully visibly (both document-state and viewer) is already implemented in the `Viewer.isVisible` method - you just need to call it at the right time. I imagine a single method:
```
Viewer.checkVisibility_() {
  if (this.isVisible()) {
   // resolve whenVisible promise
  }
  this.visibilityObservable_.fire();
}
```

And this method will need to be called: 1) In the constructor, 2) When the DocumentState's visibility changes, and 3) When Viewer's visibility changes.